crate/memories/src/batching_layer/memory.rs

```diff
@@ -0,0 +1,217 @@
+            }
+            MemoryInput::Write((guard, bindings)) => {
+                let (sender, receiver) = oneshot::channel();
+                (
+                    Operation::Write(((guard, bindings), sender)),
+                    OperationResultReceiver::<M>::Write(receiver),
+                )
+            }
+        };
+
+        if let Some(ops) = self.buffer.push(operation)? {
+            self.manage(ops).await?;
+        }
+
+        Ok(match receiver {
+            OperationResultReceiver::Read(receiver) => {
+                let result = receiver.await?.map_err(MemoryBatcherError::Memory)?;
+                MemoryOutput::Read(result)
+            }
+            OperationResultReceiver::Write(receiver) => {
+                let result = receiver.await?.map_err(MemoryBatcherError::Memory)?;
+                MemoryOutput::Write(result)
+            }
+        })
+    }
+
+    async fn manage(&self, ops: PendingOperations<M>) -> Result<(), MemoryBatcherError<M>> {
+        // Assumes the vector is homogeneous, i.e. all operations are of the same type.
+        // This should be guaranteed by the buffer.
+        match ops[0] {
+            Operation::Read(_) => {
+                // Build combined address list while tracking which addresses belong to which
+                // batch.
+                let all_addresses: Vec<_> = ops
+                    .iter()
+                    .flat_map(|op| match op {
+                        Operation::Read((addresses, _)) => addresses.clone(),
+                        _ => unreachable!(
+                            "Expected all operations to be reads, reaching this statement means \
+                             the buffer has implementation flaws at the push level."
+                        ),
+                    })
+                    .collect();
+
+                let mut aggregated_reads_results = self
+                    .inner
+                    .batch_read(all_addresses)
+                    .await
+                    .map_err(MemoryBatcherError::Memory)?;
+
+                // Distribute results to each batch's sender.
+                for (input_addresses, sender) in ops
+                    .into_iter()
+                    .map(|op| match op {
+                        Operation::Read((addresses, sender)) => (addresses, sender),
+                        _ => unreachable!(
+                            "Expected all operations to be reads, reaching this statement means \
+                             the buffer has implementation flaws at the push level."
+                        ),
+                    })
+                    .rev()
+                {
+                    let split_point = aggregated_reads_results.len() - input_addresses.len(); // This is the point where the last batch's results start.
+                    let batch_results = aggregated_reads_results.split_off(split_point); // After this call, all_results will be left containing the elements [0, split_point).
```

Comment:
Shorter names allow for clearer code:
```suggestion
                    let batch_words = words.split_off(words.len() - batch_addresses.len());
```

Reply:
Okay for **words** but I honestly had trouble understanding what `batch_addresses` mean so I kept "input"
I also had to delete the comment above as you suggested